File: frontend/lib/data/models/expense_category.dart
import 'package:flutter/material.dart';

enum ExpenseCategory {
  inventory('inventory', Icons.inventory_2_outlined),
  utilities('utilities', Icons.bolt_outlined),
  salary('salary', Icons.badge_outlined),
  transportation('transportation', Icons.local_shipping_outlined),
  rent('rent', Icons.storefront_outlined),
  supplies('supplies', Icons.shopping_basket_outlined),
  repairs('repairs', Icons.build_outlined),
  taxesPermits('taxes_permits', Icons.receipt_long_outlined),
  other('other', Icons.more_horiz_rounded);

  const ExpenseCategory(this.wireName, this.icon);

  final String wireName;
  final IconData icon;

  static ExpenseCategory parse(String? raw) => values.firstWhere(
    (category) => category.wireName == raw,
    orElse: () => ExpenseCategory.other,
  );
}

File: backend/src/features/expenses/expense_category.rs
use serde::{Deserialize, Serialize};

use crate::shared::ApiError;

/// A closed list keeps reporting meaningful. "Other" exists so nobody is ever
/// blocked from recording a real expense.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ExpenseCategory {
    Inventory,
    Utilities,
    Salary,
    Transportation,
    Rent,
    Supplies,
    Repairs,
    TaxesPermits,
    Other,
}

impl ExpenseCategory {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Inventory => "inventory",
            Self::Utilities => "utilities",
            Self::Salary => "salary",
            Self::Transportation => "transportation",
            Self::Rent => "rent",
            Self::Supplies => "supplies",
            Self::Repairs => "repairs",
            Self::TaxesPermits => "taxes_permits",
            Self::Other => "other",
        }
    }

    pub fn parse(raw: &str) -> Result<Self, ApiError> {
        match raw.trim().to_lowercase().as_str() {
            "inventory" => Ok(Self::Inventory),
            "utilities" => Ok(Self::Utilities),
            "salary" => Ok(Self::Salary),
            "transportation" => Ok(Self::Transportation),
            "rent" => Ok(Self::Rent),
            "supplies" => Ok(Self::Supplies),
            "repairs" => Ok(Self::Repairs),
            "taxes_permits" => Ok(Self::TaxesPermits),
            "other" => Ok(Self::Other),
            unknown => Err(ApiError::BadRequest(format!(
                "unknown expense category '{unknown}'"
            ))),
        }
    }

    pub fn all() -> &'static [ExpenseCategory] {
        &[
            Self::Inventory,
            Self::Utilities,
            Self::Salary,
            Self::Transportation,
            Self::Rent,
            Self::Supplies,
            Self::Repairs,
            Self::TaxesPermits,
            Self::Other,
        ]
    }
}
